source/common/network/listener_impl.cc

```diff
@@ -23,10 +23,15 @@ Address::InstanceConstSharedPtr ListenerImpl::getLocalAddress(int fd) {
 void ListenerImpl::listenCallback(evconnlistener*, evutil_socket_t fd, sockaddr* remote_addr,
                                   int remote_addr_len, void* arg) {
   ListenerImpl* listener = static_cast<ListenerImpl*>(arg);
+
+  // Create the IoSocketHandle for the fd here.
+  IoHandlePtr io_handle = std::make_unique<IoSocketHandle>(fd);
```

Comment:
Regarding address_impl.cc: I added io_socket_handle_impl.h to the hdrs section in BUILD and removed the include in address_impl.cc and it does not compile, even with io_socket_handle_lib still in the dips. Seems odd. That file only compiles if I leave the include of io_socket_handle_impl.h in. I don't understand why.

Reply:
I think the problem is you're just swapping the include.
It looks like you can refer to the pure interface in the .h file and the impl in the cc file.   What do you think?

Reply:
Sorry, I'm not sure I understand. What I did was add io_socket_handle_lib to both listen_socket_lib and listener_lib in the BUILD file, as well as removing io_socket_handle_lib.h to the hdrs section for listen_socket_lib- both compile. For address_lib, I removed the include in address_impl.cc -> compile failed. I then added io_socket_handle_impl.h to the hdrs section for address_lib (it already had io_socket_handle_lib in the deps)- still didn't compile. The only was it compiles is to have io_socket_handle_lib in the deps, and have the include of io_socket_handle_impl.h in address_impl.cc. I don't see any swapping of the include- am I missing your point?

Reply:
Matt asked you take the  io_socket_handle_impl.h out of address_impl.h.  I aagree you need it in address_impl.cc but it looks like it is still in address_impl.h

Reply:
Ah, yes. I forgot to remove that include in address_impl.h. I just removed it and added it to the hdrs section for address_lib. Even with that change listener_impl.cc doesn't compile- it seems to need address_impl.h to include io_socket_handle_impl.h. I had forgotten all these issues I encountered a few months ago. I really don't understand these dependencies.

Reply:
to be clear, you tried including io_socket_handle_impl.h in address_impl.cc and including
io_handle.h in address_impl.h?

If so please pastebin the compile error or just upload the broken patch and I'll take a look at the CI logs.

Reply:
What I did was removed io_handle.h and io_socket_handle_impl.h as includes from address_impl.h, with the changes already described in the BUILD file. When I removed the include of io_socket_handle_impl.h, I get this error:

source/common/network/address_impl.cc:162:44: error: use of undeclared identifier 'IoSocketHandle'
  IoHandlePtr io_handle = std::make_unique<IoSocketHandle>(::socket(domain, flags, 0));

Reply:
So as said above, if you remove the io_handle.h from address_impl.h you will need to add an include  io_handle.h to address_impl.cc

The .h file refers to the abstract interface, so only needs the interface file.
The .cc file creates a concrete implementation and needs the include of the impl.

It is helpful to restrict the impl to the cc file so that other files which might include address_impl.h don't get the transitive include.   Make sense?

Reply:
Yep, makes sense, thanks.